### PR TITLE
Feature MeanAveragePrecision 🔥 

### PR DESCRIPTION
**Pull request by hardikdava**
# Description

This pr is related to feature #206 and #232.

## Type of change

-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Google colab for testing:https://colab.research.google.com/drive/1MT5wc6H8WPshQfPoQITr6mLOcfd47x5E?usp=sharing

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [x] Docs updated at `docs/supervision/metrics/detection`

## Basline Metrics for comparision:

- Ultralytics
- Torchmetrics (utilizes CocoApi)

### CODE REVIEW

**Comment by hardikdava** [comment]
@SkalskiP @mayankagarwals , this feature is not completed yet, but just to keep track. I am still facing some issue in result difference to other baselines method. I will let you guys know once it is ready to tests.

**Comment by SkalskiP** [comment]
No rush, @hardikdava! In the meantime, I'm putting some work into: https://github.com/roboflow/supervision/pull/170.

**Comment by hardikdava** [comment]
Comparision between different baselines:

model=yolov8n.pt
| Framework | map50 | map75 | map |
|--|--| --|--|
| Ultralytics | 0.605 | - | 0.44
| Torchmetrics | 0.6145 | 0.4786  | 0.4473
| Supervision| 0.56066 |  0.48054 | 0.432033


@mayankagarwals  there is a bug, but I am not able to find it. Let me know if you have previous expirience with `mAP`.

**Comment by mayankagarwals** [comment]
I do have some experience but it's been a while. Can you give me a colab to reproduce these numbers for supervision? I'll take a look

**Comment by hardikdava** [comment]
> I do have some experience but it's been a while. Can you give me a colab to reproduce these numbers for supervision? I'll take a look

The link of colab is attached in the description.

**Comment by mayankagarwals** [comment]
> The link of colab is attached in the description.

Amazing, thanks!

**Comment by hardikdava** [comment]
@SkalskiP @mayankagarwals the bug is fixed. We have a working code for `mAP`.

**Comment by hardikdava** [comment]
@SkalskiP Should we add functions of plotting curves? or just values are fine?

**Comment by hardikdava** [comment]
Final report on values:

model=yolov8n.pt
| Framework | map50 | map75 | map |
|--|--| --|--|
| Ultralytics | 0.605 | - | 0.44
| Torchmetrics | 0.6145 | 0.4786  | 0.4473
| Supervision| 0.6129|  0.4807 | 0.448


model=yolov8s.pt
| Framework | map50 | map75 | map |
|--|--| --|--|
| Ultralytics | 0.759 | - | 0.585
| Torchmetrics | 0.7573 | 0.6331  | 0.5820
| Supervision| 0.7577 |  0.63638 | 0.585

File: supervision/metrics/detection.py [hunk 1]
```diff
@@ -456,3 +462,373 @@ def plot(
...
+                average_precisions.mean(1),
+            )
+            map50, map75, map = ap50.mean(), ap75.mean(), average_precisions.mean()
+
+        return cls(
+            map=map, map50=map50, map75=map75, average_precisions=average_precisions
+        )
+
+    @staticmethod
+    def detections_to_tensor(detections: Detections) -> np.ndarray:
```
Comment (SkalskiP) [inline supervision/metrics/detection.py:685]: @hardikdava, is there any reason you decided to create your implementation of `detections_to_tensor`? I'm not sure, but it looks like we could make `ConfusionMatrix.detections_to_tensor` a reusable utility. And use it in both places.

File: supervision/metrics/detection.py [hunk 2]
```diff
@@ -456,3 +462,373 @@ def plot(
...
+                if x[0].shape[0] > 1:
+                    matches = matches[matches[:, 2].argsort()[::-1]]
+                    matches = matches[np.unique(matches[:, 1], return_index=True)[1]]
+                    matches = matches[np.unique(matches[:, 0], return_index=True)[1]]
+                correct[matches[:, 1].astype(int), i] = True
+                correct[matches[:, 1].astype(int), i] = True
+        return correct
+
+    @staticmethod
+    def compute_average_precision(recall, precision):
```
Comment (SkalskiP) [inline supervision/metrics/detection.py:735]: Please add input and output types.

File: supervision/metrics/detection.py [hunk 3]
```diff
@@ -456,3 +462,373 @@ def plot(
...
+        """Compute the average precision using 101-point interpolation (COCO), given the recall and precision curves
```
Comment (SkalskiP) [inline supervision/metrics/detection.py:736]: And try to format the docstring better. It breaks documentation right now.

File: supervision/metrics/detection.py [hunk 4]
```diff
@@ -456,3 +462,373 @@ def plot(
...
+            if predictions[0].shape[1] != 6:
+                raise ValueError(
+                    f"Predictions must have shape (N, 6). Got {predictions[0].shape} instead."
+                )
+            if targets[0].shape[1] != 5:
+                raise ValueError(
+                    f"Targets must have shape (N, 5). Got {targets[0].shape} instead."
+                )
+
+    def to_dict(self):
```
Comment (SkalskiP) [inline supervision/metrics/detection.py:829]: `MeanAveragePrecision` is `dataclass`. You don't need to implement `to_dict`. It is already implemented by `dataclasses` package. Take a look [here](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict). I'd drop it.

File: supervision/metrics/detection.py [hunk 5]
```diff
@@ -456,3 +462,373 @@ def plot(
...
+                average_precisions[
+                    ci, j
+                ] = MeanAveragePrecision.compute_average_precision(
+                    recall[:, j], precision[:, j]
+                )
+
+        return average_precisions
+
+    @classmethod
+    def _validate_input_tensors(
```
Comment (SkalskiP) [inline supervision/metrics/detection.py:803]: Why not make `ConfusionMatrix._validate_input_tensors` a reusable function?

**Review by SkalskiP — CHANGES_REQUESTED** [review]

File: supervision/metrics/detection.py (hunk 3 above)
Comment (hardikdava) [reply supervision/metrics/detection.py:737]: Fixed

File: supervision/metrics/detection.py (hunk 1 above)
Comment (hardikdava) [reply supervision/metrics/detection.py]: Fixed. I had some issue with the results, so for debuging I created a new function. But I am now reusing it from Confusionmatrix.

**Comment by hardikdava** [comment]
@SkalskiP I have fixed the changes you requested.

**Comment by hardikdava** [comment]
@SkalskiP everything good? merging it?

File: supervision/metrics/detection.py [hunk 6]
```diff
@@ -12,6 +12,65 @@
...
+) -> np.ndarray:
+    """
+    Convert Supervision Detections to numpy tensors for further computation
+    Args:
+        detections (sv.Detections): Detections/Targets in the format of sv.Detections
+        with_confidence (bool): Whether to include confidence in the tensor
+    Returns:
+        (np.ndarray): Detections as numpy tensors as in (xyxy, class_id, confidence) order
+    """
+    if len(detections) == 0:
```
[the remark this answers is not in the document]
Comment (hardikdava) [reply supervision/metrics/detection.py]: @SkalskiP I wasn't able to find it. Have you already fixed it?

[the remark this answers is not in the document]
Comment (SkalskiP) [reply supervision/metrics/detection.py]: Yup.

**Comment by SkalskiP** [comment]
@hardikdava I'm working on unit tests. We will merge it when I'm done.

File: supervision/metrics/detection.py [hunk 7]
```diff
@@ -456,3 +460,329 @@ def plot(
...
+@dataclass(frozen=True)
+class MeanAveragePrecision:
+    """
+    Mean Average Precision for object detection tasks.
+
+    Attributes:
+        map (float): mAP value.
+        map50 (float): mAP value at IoU `threshold = 0.5`.
+        map75 (float): mAP value at IoU `threshold = 0.75`.
+        average_precisions (np.ndarray): values for every classes.
```
Comment (SkalskiP) [inline supervision/metrics/detection.py:474]: @hardikdava could you help me out here? What is stored under the `average_precisions` field?

Comment (hardikdava) [reply supervision/metrics/detection.py:474]: Average precision holds the result of ap at different iou threshold starting from 0.5 to 0.95 at an interval of 0.5 also the results are for each class. Columns represents classes and rows represents iou threshold.

Comment (SkalskiP) [reply supervision/metrics/detection.py:474]: So what would be the expected shape of this np array? (N, 10)? Where N is the amount of classes?

Comment (hardikdava) [reply supervision/metrics/detection.py:474]: Yeah, numpy array of (num of class, 10).

Comment (SkalskiP) [reply supervision/metrics/detection.py:474]: I Done some unit teats and it looks like the shape is different :/

Comment (SkalskiP) [reply supervision/metrics/detection.py:474]: ### Test 1

```python
>>> import supervision as sv
>>> import numpy as np

>>> predictions = [
...     np.empty((0, 6), dtype=np.float32),
... ]
>>> targets = [
...     np.empty((0, 5), dtype=np.float32),
... ]

>>> sv.MeanAveragePrecision.from_tensors(predictions=predictions, targets=targets)
# MeanAveragePrecision(map=0, map50=0, map75=0, average_precisions=[])
```

### Test 2

```python
>>> import supervision as sv
>>> import numpy as np

>>> predictions = [
...     np.array([
...         [0.0, 0.0, 1.0, 1.0, 1, 0.9],
...     ]),
... ]
>>> targets = [
...     np.array([
...         [0.0, 0.0, 1.0, 1.0, 1],
...     ]),
... ]

>>> sv.MeanAveragePrecision.from_tensors(predictions=predictions, targets=targets)
# MeanAveragePrecision(map=0.9949999999999999, map50=0.995, map75=0.995, average_precisions=array([      0.995]))
```

### Test 3

```python
>>> import supervision as sv
>>> import numpy as np

>>> predictions = [
...     np.array([
...         [0.0, 0.0, 1.0, 1.0, 1, 0.9],
...     ]),
...     np.array([
...         [0.0, 0.0, 1.0, 1.0, 1, 0.9],
...     ]),
...     np.array([
...         [0.0, 0.0, 1.0, 1.0, 2, 0.9],
...     ]),
... ]
>>> targets = [
...     np.array([
...         [0.0, 0.0, 1.0, 1.0, 1],
...     ]),
...     np.array([
...         [0.0, 0.0, 0.8, 0.8, 1, 0.9],
...     ]),
...     np.array([
...         [0.0, 0.0, 1.0, 1.0, 2, 0.9],
...     ]),
... ]

>>> result = sv.MeanAveragePrecision.from_tensors(predictions=predictions, targets=targets)
# MeanAveragePrecision(map=0.864625, map50=0.995, map75=0.80875, average_precisions=array([    0.73425,       0.995]))

```

Comment (SkalskiP) [reply supervision/metrics/detection.py:474]: Those are simply average precisions per class. Not per threshold but already averaged.

@hardikdava is it what you intended?

Comment (hardikdava) [reply supervision/metrics/detection.py:474]: Ohh yeah, average_precision is already per class map values. Sorry for miscommunication. This is intended.

**Pull request merged**